### PR TITLE
add IrohaConsumer implementation

### PR DESCRIPTION
**Pull request by Alexey-N-Chernyshov**
Implementation of IrohaConsumerImpl
Add protobuf to gradle build script

### CODE REVIEW

File: src/main/kotlin/sideChain/iroha/IrohaConsumerImpl.kt [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+import io.grpc.ManagedChannelBuilder
+
+/**
+ * Dummy implementation of [IrohaConsumer] with effective dependency
+ */
+class IrohaConsumerImpl(notary: Notary) : IrohaConsumer {
+
+    /** Iroha keypair */
+    val keypair: Keypair
```
Comment (muratovv) [inline src/main/kotlin/sideChain/iroha/IrohaConsumerImpl.kt]: This key pair should be set into ctor.

File: src/main/kotlin/sideChain/iroha/IrohaConsumerImpl.kt [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+    init {
+        notary.irohaOutput().subscribe {
+            onIrohaEvent(it)
+        }
+        val crypto = ModelCrypto()
+        keypair = crypto.convertFromExisting(
```
Comment (muratovv) [inline src/main/kotlin/sideChain/iroha/IrohaConsumerImpl.kt]: This is a use case. We have to move it into initialization Iroha class.

File: src/main/kotlin/sideChain/iroha/IrohaConsumerImpl.kt [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+            readKeyFromFile(Configs.pubkeyPath),
+            readKeyFromFile(Configs.privkeyPath)
+        )
+    }
+
+    /**
+     * Read key from file specified by [path]
+     */
+    private fun readKeyFromFile(path: String): String {
```
Comment (muratovv) [inline src/main/kotlin/sideChain/iroha/IrohaConsumerImpl.kt]: sab

File: src/main/kotlin/sideChain/iroha/IrohaConsumerImpl.kt [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+            return String(java.nio.file.Files.readAllBytes(Paths.get(path)))
+        } catch (e: java.io.IOException) {
+            logger.error("Unable to read key files.\n $e")
+            System.exit(1)
+        }
+        return ""
+    }
+
+    /** Convert [ByteVector] to [ByteArray] */
+    private fun toByteArray(blob: ByteVector): ByteArray {
```
Comment (muratovv) [inline src/main/kotlin/sideChain/iroha/IrohaConsumerImpl.kt]: This function looks like Util. Maybe remove it from consumer code and use a dependency?

File: src/main/kotlin/sideChain/iroha/IrohaConsumerImpl.kt [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        val size = blob.size().toInt()
+        val bs = ByteArray(size)
+        for (i in 0 until size) {
+            bs[i] = blob.get(i).toByte()
+        }
+        return bs
+    }
+
+    /** Add [commands] to transaction builder */
```
Comment (muratovv) [inline src/main/kotlin/sideChain/iroha/IrohaConsumerImpl.kt]: The documentation misses parameters

File: src/main/kotlin/sideChain/iroha/IrohaConsumerImpl.kt [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    private fun addCommandToBuilder(
```
Comment (muratovv) [inline src/main/kotlin/sideChain/iroha/IrohaConsumerImpl.kt]: We have to move convertors from Notary model to Iroha as a dependency and set it into ctor

File: src/main/kotlin/sideChain/iroha/IrohaConsumerImpl.kt [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+            var protoTx: BlockOuterClass.Transaction? = null
+            try {
+                protoTx = BlockOuterClass.Transaction.parseFrom(bs)
+            } catch (e: InvalidProtocolBufferException) {
+                logger.error { "Exception while converting byte array to protobuf: ${e.message}" }
+                System.exit(1)
+            }
+
+            // Send transaction to iroha
+            val channel =
```
Comment (muratovv) [inline src/main/kotlin/sideChain/iroha/IrohaConsumerImpl.kt]: I think we have to move network to own level

File: src/main/kotlin/main/Configs.kt [hunk 8]
```diff
@@ -0,0 +1,25 @@
+package main
+
+/**
+ * Class incapsulates all configuration parameters for the application.
+ */
+object Configs {
+
+    /** Port for refund REST API */
+    var refundPort = 8080
```
[the remark this answers is not in the document]
Comment (Alexey-N-Chernyshov) [reply src/main/kotlin/main/Configs.kt]: To change it with another from CLI arguments or load from file. I suggest to replace the class with https://github.com/npryce/konfig or smth similar if we need it.

File: example/admin@test.pub [hunk 9]
```diff
@@ -0,0 +1 @@
+407e57f50ca48969b08ba948171bb2435e035d82cec417e18e4a38f5fb113f83
```
[the remark this answers is not in the document]
Comment (Alexey-N-Chernyshov) [reply example/admin@test.pub]: Task named "initialisation of notary" should handle it.

File: src/main/kotlin/sideChain/iroha/IrohaConsumerImpl.kt (hunk 4 above)
Comment (Alexey-N-Chernyshov) [reply src/main/kotlin/sideChain/iroha/IrohaConsumerImpl.kt]: I'm confused with `signAndAddSignature()`. It returns custom type `ByteVector ` that should be converted to `ByteArray`. I don't think that function can be used elsewhere.

File: src/main/kotlin/notary/IrohaOrderedBatch.kt [hunk 10]
```diff
@@ -1,15 +1,14 @@
...
-    /** Batch is ordered container of transaction, each can contain several [IrohaCommand] */
-    var transactions = ArrayList<Array<IrohaCommand>>()
+    /** Batch is ordered container of transaction */
+    val transactions: Array<out IrohaTransaction>
 
-    /** Add transaction to the batch */
-    fun addTransaction(transaction: Array<IrohaCommand>) {
-        transactions.add(transaction)
+    init {
+        transactions = txs
```
Comment (muratovv) [inline src/main/kotlin/notary/IrohaOrderedBatch.kt]: Init should be omitted with trivial intialization in ctor.

File: src/main/kotlin/notary/IrohaOrderedBatch.kt [hunk 11]
```diff
@@ -1,15 +1,14 @@
...
- * Class represents [Notary] intention to [sideChain.iroha.IrohaConsumer]
+ * Class represents [Notary] intention to [sideChain.iroha.consumer.IrohaConsumer] to add batch transaction
  */
-class IrohaOrderedBatch {
+class IrohaOrderedBatch(vararg txs: IrohaTransaction) {
 
-    /** Batch is ordered container of transaction, each can contain several [IrohaCommand] */
-    var transactions = ArrayList<Array<IrohaCommand>>()
+    /** Batch is ordered container of transaction */
+    val transactions: Array<out IrohaTransaction>
```
Comment (muratovv) [inline src/main/kotlin/notary/IrohaOrderedBatch.kt]: Rework Array with List

File: src/main/kotlin/notary/NotaryInitialization.kt [hunk 12]
```diff
@@ -71,7 +74,8 @@ class NotaryInitialization {
      */
     fun initIrohaConsumer() {
         logger.info { "Init Iroha consumer" }
-        irohaConsumer = IrohaConsumerImpl(notary)
+        val cryptoProvider = IrohaCryptoProviderImpl(Configs.pubkeyPath, Configs.privkeyPath)
```
Comment (muratovv) [inline src/main/kotlin/notary/NotaryInitialization.kt]: IrohaCryptoProvide has to get own dependencies e.g. keys directly. Code that loading keys from files should be used in initialization class explicitly.

File: src/main/kotlin/notary/NotaryStub.kt [hunk 13]
```diff
@@ -14,16 +13,22 @@ class NotaryStub(
...
     override fun onIrohaEvent(irohaEvent: NotaryEvent.IrohaChainEvent) {
         logger.info { "Notary performs IROHA event" }
     }
 
     /**
-     * Dummy implementation with relaying eth events for output
+     * Relay side chain [NotaryEvent] to Iroha output
      */
     override fun irohaOutput(): Observable<IrohaOrderedBatch> {
         return io.reactivex.Observable.merge(
```
Comment (muratovv) [inline src/main/kotlin/notary/NotaryStub.kt]: Make a TODO with moving merge into ctor.

File: src/main/kotlin/sideChain/iroha/IrohaInitializtion.kt [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+class IrohaInitializtion {
+
+    /**
+     * Load iroha binding library
+     */
+    fun loadIrohaLibrary() {
+        try {
+            System.loadLibrary("irohajava")
+        } catch (e: UnsatisfiedLinkError) {
+            System.err.println("Native code library failed to load. \n" + e)
```
Comment (muratovv) [inline src/main/kotlin/sideChain/iroha/IrohaInitializtion.kt]: Use logger instead

File: src/main/kotlin/sideChain/iroha/consumer/IrohaConsumerImpl.kt [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+    val keypair = cryptoProvider.keypair
+
+    init {
+        source.subscribe {
+            irohaOutput(it)
+        }
+    }
+
+    /** Convert [ByteVector] to [ByteArray] */
+    private fun toByteArray(blob: ByteVector): ByteArray {
```
Comment (muratovv) [inline src/main/kotlin/sideChain/iroha/consumer/IrohaConsumerImpl.kt]: Move this and further method to own Iroha Util package for improving testability.

File: src/main/kotlin/sideChain/iroha/consumer/IrohaConsumerImpl.kt [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+        }
+        return protoTx
+    }
+
+    /** Callback to be called to write to Iroha */
+    override fun irohaOutput(batch: IrohaOrderedBatch) {
+        IrohaNetworkImpl.logger.info { "TX to IROHA" }
+
+        // TODO rework with batch transactions
+        for (tx in batch.transactions) {
```
Comment (muratovv) [inline src/main/kotlin/sideChain/iroha/consumer/IrohaConsumerImpl.kt]: I think here in output we may provide observable with protobuf objects for the transport layer

File: src/main/kotlin/sideChain/iroha/consumer/IrohaCryptoProviderImpl.kt [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+        keypair = crypto.convertFromExisting(
+            readKeyFromFile(pubkeyPath),
+            readKeyFromFile(privkeyPath)
+        )
+    }
+
+    /**
+     * Read key from file specified by [path]
+     */
+    private fun readKeyFromFile(path: String): String {
```
Comment (muratovv) [inline src/main/kotlin/sideChain/iroha/consumer/IrohaCryptoProviderImpl.kt]: This method doesn't relate to keys in terms of types. We have to move it into util package also. And use it here as a dependency.

File: src/main/kotlin/sideChain/iroha/consumer/IrohaCryptoProviderImpl.kt [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+        try {
+            return String(java.nio.file.Files.readAllBytes(Paths.get(path)))
+        } catch (e: java.io.IOException) {
+            IrohaConsumerImpl.logger.error { "Unable to read key files.\n $e" }
+            System.exit(1)
+        }
+        return ""
```
Comment (muratovv) [inline src/main/kotlin/sideChain/iroha/consumer/IrohaCryptoProviderImpl.kt]: Also, type is not suitable. Better to use the representation of the result type here.

File: src/main/kotlin/sideChain/iroha/consumer/IrohaCryptoProviderImpl.kt [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+
+    /**
+     * Read key from file specified by [path]
+     */
+    private fun readKeyFromFile(path: String): String {
+        try {
+            return String(java.nio.file.Files.readAllBytes(Paths.get(path)))
+        } catch (e: java.io.IOException) {
+            IrohaConsumerImpl.logger.error { "Unable to read key files.\n $e" }
+            System.exit(1)
```
Comment (muratovv) [inline src/main/kotlin/sideChain/iroha/consumer/IrohaCryptoProviderImpl.kt]: Why this method implicit provide that behaviour? This has to be configured for use in consumer class instead.

File: src/main/kotlin/sideChain/iroha/consumer/IrohaConsumerImpl.kt [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+        // Convert ByteVector to byte array
+        val bs = toByteArray(txblob)
+
+        // create proto object
+        lateinit var protoTx: BlockOuterClass.Transaction
+        try {
+            protoTx = BlockOuterClass.Transaction.parseFrom(bs)
+        } catch (e: InvalidProtocolBufferException) {
+            logger.error { "Exception while converting byte array to protobuf: ${e.message}" }
+            System.exit(1)
```
Comment (muratovv) [inline src/main/kotlin/sideChain/iroha/consumer/IrohaConsumerImpl.kt]: This is not expected behaviour, think about problems in terms of types.

File: src/main/kotlin/sideChain/iroha/consumer/IrohaConsumerImpl.kt [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+        }
+        return protoTx
+    }
+
+    /** Callback to be called to write to Iroha */
+    override fun irohaOutput(batch: IrohaOrderedBatch) {
+        IrohaNetworkImpl.logger.info { "TX to IROHA" }
```
Comment (muratovv) [inline src/main/kotlin/sideChain/iroha/consumer/IrohaConsumerImpl.kt]: I guess better to stay logging with "TX to IROHA" only in the transport layer.

File: src/main/kotlin/sideChain/iroha/consumer/IrohaNetworkImpl.kt [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+
+    /**
+     * Send transaction to iroha
+     * @param protoTx protobuf representation of transaction
+     */
+    override fun send(protoTx: BlockOuterClass.Transaction) {
+        logger.info { "TX to IROHA" }
+
+        // Send transaction to iroha
+        val channel =
```
Comment (muratovv) [inline src/main/kotlin/sideChain/iroha/consumer/IrohaNetworkImpl.kt]: Why we configure client on each send call? Better to move it into ctor.

**Pull request merged**